### PR TITLE
enhance: Tidy compactor and remove dup codes

### PR DESCRIPTION
**Pull request by XuanYang-cn**
See also: #32451

### CODE REVIEW

**Comment by XuanYang-cn** [comment]
before:
fu7suNSFSF
after:
EHamAMELpD

**Comment by XuanYang-cn** [comment]
image

**Comment by XuanYang-cn** [comment]
see also: #32448

**Comment by XuanYang-cn** [comment]
rerun ut

**Comment by XuanYang-cn** [comment]
rerun ut

File: internal/datanode/compaction/segment_writer.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+func (w *SegmentWriter) Finish(actualRowCount int64) (*storage.Blob, error) {
+	w.writer.Flush()
+	codec := storage.NewInsertCodecWithSchema(&etcdpb.CollectionMeta{ID: w.collectionID, Schema: w.sch})
+	return codec.SerializePkStats(w.pkstats, actualRowCount)
+}
+
+func (w *SegmentWriter) IsFull() bool {
+	w.writer.Flush()
```
Comment (bigsheeper) [inline internal/datanode/compaction/segment_writer.go:77]: Why do we need to Flush each time we check IsFull?

File: internal/datanode/compaction_executor.go [hunk 2]
```diff
@@ -34,10 +35,10 @@ const (
 )
 
 type compactionExecutor struct {
```
Comment (bigsheeper) [inline internal/datanode/compaction_executor.go:37]: also move compactionExecutor to datanode/compaction/ dir?

File: internal/datanode/l0_compactor.go [hunk 3]
```diff
@@ -50,7 +52,7 @@ import (
 )
 
 type levelZeroCompactionTask struct {
```
Comment (bigsheeper) [inline internal/datanode/l0_compactor.go:54]: also move l0_compactor.go to datanode/compaction/ dir?

File: internal/datanode/compaction_executor.go (hunk 2 above)
Comment (XuanYang-cn) [reply internal/datanode/compaction_executor.go:37]: Too many changes in one PR, maybe you could move it inside.

File: internal/datanode/l0_compactor.go (hunk 3 above)
Comment (XuanYang-cn) [reply internal/datanode/l0_compactor.go:54]: YES, planning to do so.

File: internal/datanode/compaction/segment_writer.go (hunk 1 above)
Comment (XuanYang-cn) [reply internal/datanode/compaction/segment_writer.go:77]: So donot check isFull too frequently, its expensive

**Comment by XuanYang-cn** [comment]
rerun ut

**Pull request merged**